File: commons/javafx3d/src/main/java/com/cedarsoft/commons/javafx/threed/SubSceneInputEventSupport.java
package com.cedarsoft.commons.javafx.threed;

import javax.annotation.Nonnull;

import com.cedarsoft.unit.other.deg;

import javafx.scene.Camera;
import javafx.scene.Node;
import javafx.scene.SubScene;
import javafx.scene.input.GestureEvent;
import javafx.scene.input.MouseEvent;
import javafx.scene.input.ScrollEvent;
import javafx.scene.transform.Rotate;
import javafx.scene.transform.Translate;

/**
 * A utility class that converts {@link MouseEvent} events and {@link ScrollEvent} events
 * into rotations of a {@link Node} which must be a child of a {@link SubScene}.
 */
public class SubSceneInputEventSupport {
  @deg
  private static final int ANGLE_FULL_CIRCLE = 360;

  private double mousePosX;
  private double mousePosY;
  private double mouseOldX;
  private double mouseOldY;

  @Nonnull 
  private final Rotate rotateX;
  @Nonnull 
  private final Rotate rotateY;
  @Nonnull 
  private final Rotate rotateZ;
  @Nonnull
  private final Translate cameraTranslate;

  public SubSceneInputEventSupport(@Nonnull SubScene subScene) {
    rotateX = new Rotate(0, Rotate.X_AXIS);
    rotateY = new Rotate(0, Rotate.Y_AXIS);
    rotateZ = new Rotate(0, Rotate.Z_AXIS);

    Camera camera = subScene.getCamera();
    cameraTranslate = new Translate(0, 0, 0);
    camera.getTransforms().add(cameraTranslate);

    subScene.setOnScroll(event -> {
      double scale = getTranslateZScale(event);
      double oldZ = cameraTranslate.getZ();
      double newZ = oldZ + event.getDeltaY() * scale;
      cameraTranslate.setZ(newZ);
    });

    subScene.setOnMousePressed(event -> {
      mousePosX = event.getSceneX();
      mousePosY = event.getSceneY();
      mouseOldX = mousePosX;
      mouseOldY = mousePosY;
      if (event.getClickCount() == 2) {
        resetAll();
      }
    });

    subScene.setOnMouseDragged(event -> {
      mouseOldX = mousePosX;
      mouseOldY = mousePosY;
      mousePosX = event.getSceneX();
      mousePosY = event.getSceneY();

      final double mouseDeltaX = mousePosX - mouseOldX;
      final double mouseDeltaY = mousePosY - mouseOldY;

      if (event.isPrimaryButtonDown()) {
        if (event.isAltDown()) { //roll
          if (Math.abs(mouseDeltaX) > Math.abs(mouseDeltaY)) {
            rotateZ.setAngle(computeNewAngle(rotateZ.getAngle(), mouseDeltaX * getRotateZScale(event)));
          }
          else {
            rotateZ.setAngle(computeNewAngle(rotateZ.getAngle(), mouseDeltaY * getRotateZScale(event)));
          }
        }
        else {
          double scale = getRotateXYScale(event);
          rotateY.setAngle(computeNewAngle(rotateY.getAngle(), mouseDeltaX * scale));
          rotateX.setAngle(computeNewAngle(rotateX.getAngle(), -mouseDeltaY * scale));
        }
      }
      else {
        double oldX = cameraTranslate.getX();
        double newX = oldX - mouseDeltaX * getTranslateXYScale(event);
        cameraTranslate.setX(newX);

        double oldY = cameraTranslate.getY();
        double newY = oldY - mouseDeltaY * getTranslateXYScale(event);
        cameraTranslate.setY(newY);
      }
    });

  }

  private static double computeNewAngle(double currentAngle, double offset) {
    return (currentAngle + offset) % ANGLE_FULL_CIRCLE;
  }

  public void install(@Nonnull Node node) {
    node.getTransforms().addAll(rotateX, rotateY, rotateZ);
  }

  public void resetRotation() {
    rotateX.setAngle(0);
    rotateY.setAngle(0);
    rotateZ.setAngle(0);
  }

  public void resetTranslation() {
    cameraTranslate.setX(0);
    cameraTranslate.setY(0);
    cameraTranslate.setZ(0);
  }

  public void resetAll() {
    resetTranslation();
    resetRotation();
  }

  @SuppressWarnings("MagicNumber")
  private static double getRotateXYScale(@Nonnull MouseEvent event) {
    double scale = 0.75;
    if (event.isControlDown()) {
      scale = 0.1;
    }
    if (event.isShiftDown()) {
      scale = 2.0;
    }
    return scale;
  }

  @SuppressWarnings("MagicNumber")
  private static double getRotateZScale(@Nonnull MouseEvent event) {
    double scale = 2.0;
    if (event.isControlDown()) {
      scale = 1.0;
    }
    if (event.isShiftDown()) {
      scale = 5.0;
    }
    return scale;
  }

  @SuppressWarnings("MagicNumber")
  private static double getTranslateXYScale(@Nonnull MouseEvent event) {
    double scale = 2.0;
    if (event.isControlDown()) {
      scale = 0.5;
    }
    if (event.isShiftDown()) {
      scale = 5.0;
    }
    return scale;
  }

  @SuppressWarnings("MagicNumber")
  private static double getTranslateZScale(@Nonnull GestureEvent event) {
    double scale = 25.0;
    if (event.isControlDown()) {
      scale = 2.5;
    }
    if (event.isShiftDown()) {
      scale = 75.0;
    }
    return scale;
  }

}

File: commons/javafx3d/src/main/kotlin/com/cedarsoft/commons/javafx/threed/SubSceneInputEventSupport.kt
package com.cedarsoft.commons.javafx.threed

import com.cedarsoft.unit.other.deg
import javafx.event.EventHandler
import javafx.scene.Node
import javafx.scene.SubScene
import javafx.scene.input.GestureEvent
import javafx.scene.input.MouseEvent
import javafx.scene.input.ScrollEvent
import javafx.scene.transform.Rotate
import javafx.scene.transform.Translate

/**
 * A utility class that converts [MouseEvent] events and [ScrollEvent] events
 * into rotations of a [Node] which must be a child of a [SubScene].
 */
class SubSceneInputEventSupport(subScene: SubScene) {
  private var mousePosX = 0.0
  private var mousePosY = 0.0
  private var mouseOldX = 0.0
  private var mouseOldY = 0.0

  private val rotateX: Rotate = Rotate(0.0, Rotate.X_AXIS)
  private val rotateY: Rotate = Rotate(0.0, Rotate.Y_AXIS)
  private val rotateZ: Rotate = Rotate(0.0, Rotate.Z_AXIS)

  private val cameraTranslate: Translate

  init {
    val camera = subScene.camera
    cameraTranslate = Translate(0.0, 0.0, 0.0)
    camera.transforms.add(cameraTranslate)
    subScene.onScroll = EventHandler { event: ScrollEvent ->
      val scale = getTranslateZScale(event)
      val oldZ = cameraTranslate.z
      val newZ = oldZ + event.deltaY * scale
      cameraTranslate.z = newZ
    }
    subScene.onMousePressed = EventHandler { event: MouseEvent ->
      mousePosX = event.sceneX
      mousePosY = event.sceneY
      mouseOldX = mousePosX
      mouseOldY = mousePosY
      if (event.clickCount == 2) {
        resetAll()
      }
    }
    subScene.onMouseDragged = EventHandler { event: MouseEvent ->
      mouseOldX = mousePosX
      mouseOldY = mousePosY
      mousePosX = event.sceneX
      mousePosY = event.sceneY
      val mouseDeltaX = mousePosX - mouseOldX
      val mouseDeltaY = mousePosY - mouseOldY
      if (event.isPrimaryButtonDown) {
        if (event.isAltDown) { //roll
          if (Math.abs(mouseDeltaX) > Math.abs(mouseDeltaY)) {
            rotateZ.angle = computeNewAngle(rotateZ.angle, mouseDeltaX * getRotateZScale(event))
          } else {
            rotateZ.angle = computeNewAngle(rotateZ.angle, mouseDeltaY * getRotateZScale(event))
          }
        } else {
          val scale = getRotateXYScale(event)
          rotateY.angle = computeNewAngle(rotateY.angle, mouseDeltaX * scale)
          rotateX.angle = computeNewAngle(rotateX.angle, -mouseDeltaY * scale)
        }
      } else {
        val oldX = cameraTranslate.x
        val newX = oldX - mouseDeltaX * getTranslateXYScale(event)
        cameraTranslate.x = newX
        val oldY = cameraTranslate.y
        val newY = oldY - mouseDeltaY * getTranslateXYScale(event)
        cameraTranslate.y = newY
      }
    }
  }

  fun install(node: Node) {
    node.transforms.addAll(rotateX, rotateY, rotateZ)
  }

  fun resetRotation() {
    rotateX.angle = 0.0
    rotateY.angle = 0.0
    rotateZ.angle = 0.0
  }

  fun resetTranslation() {
    cameraTranslate.x = 0.0
    cameraTranslate.y = 0.0
    cameraTranslate.z = 0.0
  }

  fun resetAll() {
    resetTranslation()
    resetRotation()
  }

  companion object {
    private const val ANGLE_FULL_CIRCLE: @deg Int = 360

    private fun computeNewAngle(currentAngle: Double, offset: Double): Double {
      return (currentAngle + offset) % ANGLE_FULL_CIRCLE
    }

    private fun getRotateXYScale(event: MouseEvent): Double {
      var scale = 0.75
      if (event.isControlDown) {
        scale = 0.1
      }
      if (event.isShiftDown) {
        scale = 2.0
      }
      return scale
    }

    private fun getRotateZScale(event: MouseEvent): Double {
      var scale = 2.0
      if (event.isControlDown) {
        scale = 1.0
      }
      if (event.isShiftDown) {
        scale = 5.0
      }
      return scale
    }

    private fun getTranslateXYScale(event: MouseEvent): Double {
      var scale = 2.0
      if (event.isControlDown) {
        scale = 0.5
      }
      if (event.isShiftDown) {
        scale = 5.0
      }
      return scale
    }

    private fun getTranslateZScale(event: GestureEvent): Double {
      var scale = 25.0
      if (event.isControlDown) {
        scale = 2.5
      }
      if (event.isShiftDown) {
        scale = 75.0
      }
      return scale
    }
  }
}
